src/jit/lclvars.h

```diff
@@ -0,0 +1,750 @@
+
+    BYTE* lvGcLayout; // GC layout info for structs
+
+#if ASSERTION_PROP
+    BlockSet   lvRefBlks;          // Set of blocks that contain refs
+    GenTreePtr lvDefStmt;          // Pointer to the statement with the single definition
+    void       lvaDisqualifyVar(); // Call to disqualify a local variable from use in optAddCopies
+#endif
+    var_types TypeGet() const
+    {
+        return (var_types)lvType;
+    }
+    bool lvStackAligned() const
+    {
+        assert(lvIsStructField);
+        return ((lvFldOffset % sizeof(void*)) == 0);
+    }
+    bool lvNormalizeOnLoad() const
+    {
+        return varTypeIsSmall(TypeGet()) &&
+               // lvIsStructField is treated the same as the aliased local, see fgDoNormalizeOnStore.
+               (lvIsParam || lvAddrExposed || lvIsStructField);
+    }
+
+    bool lvNormalizeOnStore()
+    {
+        return varTypeIsSmall(TypeGet()) &&
+               // lvIsStructField is treated the same as the aliased local, see fgDoNormalizeOnStore.
+               !(lvIsParam || lvAddrExposed || lvIsStructField);
+    }
+
+    void lvaResetSortAgainFlag(Compiler* pComp);
+    void decRefCnts(BasicBlock::weight_t weight, Compiler* pComp, bool propagate = true);
+    void incRefCnts(BasicBlock::weight_t weight, Compiler* pComp, bool propagate = true);
+    void setPrefReg(regNumber regNum, Compiler* pComp);
+    void addPrefReg(regMaskTP regMask, Compiler* pComp);
+    bool IsFloatRegType() const
+    {
+        return isFloatRegType(lvType) || lvIsHfaRegArg();
+    }
+    var_types GetHfaType() const
+    {
+        return lvIsHfa() ? (lvHfaTypeIsFloat() ? TYP_FLOAT : TYP_DOUBLE) : TYP_UNDEF;
+    }
+    void SetHfaType(var_types type)
+    {
+        assert(varTypeIsFloating(type));
+        lvSetHfaTypeIsFloat(type == TYP_FLOAT);
+    }
+
+#ifndef LEGACY_BACKEND
+    var_types lvaArgType();
+#endif
+
+    PerSsaArray lvPerSsaData;
+
+#ifdef DEBUG
+    // Keep track of the # of SsaNames, for a bounds check.
+    unsigned lvNumSsaNames;
+#endif
+
+    // Returns the address of the per-Ssa data for the given ssaNum (which is required
+    // not to be the SsaConfig::RESERVED_SSA_NUM, which indicates that the variable is
+    // not an SSA variable).
+    LclSsaVarDsc* GetPerSsaData(unsigned ssaNum)
+    {
+        assert(ssaNum != SsaConfig::RESERVED_SSA_NUM);
+        assert(SsaConfig::RESERVED_SSA_NUM == 0);
+        unsigned zeroBased = ssaNum - SsaConfig::UNINIT_SSA_NUM;
+        assert(zeroBased < lvNumSsaNames);
+        return &lvPerSsaData.GetRef(zeroBased);
+    }
+
+#ifdef DEBUG
+public:
+    void PrintVarReg() const
+    {
+        if (isRegPairType(TypeGet()))
+        {
+            printf("%s:%s", getRegName(lvOtherReg), // hi32
+                   getRegName(lvRegNum));           // lo32
+        }
+        else
+        {
+            printf("%s", getRegName(lvRegNum));
+        }
+    }
+#endif // DEBUG
+
+}; // class LclVarDsc
+
+class LclVarTable final
+{
```

Comment:
> I'm not fond of all the references. There are a zillion functions in the JIT that take a `Compiler*`; I don't see why these need to take a `Compiler&` instead.

Defensive coding: it's much, much more difficult to accidentally manufacture a null `&` than a null `*`. If you feel strongly about this, though, I'm happy to change these to `Compiler*`s.

>  It also seems like `operator[]` should return a `LclVarDsc*`.

That's two votes for that option (@mikedn) mentioned the same. I'll make the change.

On a related topic, I had suggested in that discussion that we could add a `LclVarDsc* operator+(unsigned index)` to `LclVarTable`, which would avoid the need to make most of the edits in these files. I shied away from that because I felt that it made the underlying behavior less clear, but I'm not entirely opposed to adding it. Thoughts?